File: AssetManager/UserInterface/MyControls/WindowList.vb
﻿Public Class WindowList
    Private WithEvents RefreshTimer As Timer
    Private MyParentForm As Form
    Private DropDownControl As ToolStripDropDownButton
    Private intFormCount As Integer
    Sub New(ParentForm As Form, DropDownCtl As ToolStripDropDownButton)
        MyParentForm = ParentForm
        DropDownControl = DropDownCtl
        DropDownControl.Visible = False
        Init()
    End Sub
    Private Sub Init()
        InitializeTimer()
    End Sub
    Private Sub InitializeTimer()
        RefreshTimer = New Timer
        RefreshTimer.Interval = 200
        RefreshTimer.Enabled = True
        AddHandler RefreshTimer.Tick, AddressOf RefreshTimer_Tick
    End Sub
    Private Sub RefreshTimer_Tick(sender As Object, e As EventArgs) Handles RefreshTimer.Tick
        If FormCount(MyParentForm) < 1 Then
            DropDownControl.Visible = False
        Else
            DropDownControl.Visible = True
        End If
        If FormCount(MyParentForm) <> intFormCount Then
            If Not DropDownControl.DropDown.Focused Then
                DropDownControl.DropDownItems.Clear()
                BuildWindowList(MyParentForm, DropDownControl.DropDownItems)
                intFormCount = FormCount(MyParentForm)
                DropDownControl.Text = CountText(intFormCount)
            End If
        End If
    End Sub
    Private Function FormCount(ParentForm As Form) As Integer
        Dim i As Integer = 0
        For Each frm As Form In My.Application.OpenForms
            If Not frm.IsDisposed And Not frm.Modal And frm IsNot ParentForm Then
                If frm.Tag Is ParentForm Then
                    i += 1
                    i += FormCount(frm)
                End If
            End If
        Next
        Return i
    End Function
    ''' <summary>
    ''' Recursively build ToolStripItemCollections of Forms and their Children and add them to the ToolStrip. Making sure to add SibiMain to the top of the list.
    ''' </summary>
    ''' <param name="ParentForm">Form to add to ToolStrip.</param>
    ''' <param name="TargetMenuItem">Item to add the Form item to.</param>
    Private Sub BuildWindowList(ParentForm As Form, ByRef TargetMenuItem As ToolStripItemCollection)
        For Each frm As Form In ListOfChilden(ParentForm)
            If HasChildren(frm) Then
                Dim NewDropDown As ToolStripMenuItem = NewMenuItem(frm)
                If TypeOf frm Is frmSibiMain Then
                    TargetMenuItem.Insert(0, NewDropDown)
                Else
                    TargetMenuItem.Add(NewDropDown)
                End If
                BuildWindowList(frm, NewDropDown.DropDownItems)
            Else
                If TypeOf frm Is frmSibiMain Then
                    TargetMenuItem.Insert(0, NewMenuItem(frm))
                Else
                    TargetMenuItem.Add(NewMenuItem(frm))
                End If
            End If
        Next
    End Sub
    Private Function NewMenuItem(frm As Form) As ToolStripMenuItem
        Dim newitem As New ToolStripMenuItem
        newitem.Text = frm.Text
        newitem.Image = frm.Icon.ToBitmap
        newitem.Tag = frm
        newitem.ToolTipText = "Right-Click to close."
        AddHandler newitem.MouseDown, AddressOf WindowClick
        Return newitem
    End Function
    Private Function HasChildren(ParentForm As Form) As Boolean
        For Each frm As Form In My.Application.OpenForms
            If frm.Tag Is ParentForm And Not frm.IsDisposed Then
                Return True
            End If
        Next
        Return False
    End Function
    Private Function ListOfChilden(ParentForm As Form) As List(Of Form)
        Dim tmpList As New List(Of Form)
        For Each frm As Form In My.Application.OpenForms
            If frm.Tag Is ParentForm And Not frm.IsDisposed Then
                tmpList.Add(frm)
            End If
        Next
        Return tmpList
    End Function
    Private Sub WindowClick(sender As Object, e As MouseEventArgs)
        Dim item As ToolStripItem = CType(sender, ToolStripItem)
        If e.Button = MouseButtons.Right Then
            Dim frm As Form = CType(item.Tag, Form)
            frm.Dispose()
            GC.Collect()
            If DropDownControl.DropDownItems.Count < 1 Then
                DropDownControl.Visible = False
                DropDownControl.DropDownItems.Clear()
                item.Dispose()
            Else
                item.Visible = False
                item.Dispose()
                DropDownControl.DropDownItems.Remove(item)
                intFormCount = FormCount(MyParentForm)
                DropDownControl.Text = CountText(intFormCount)
            End If
        ElseIf e.Button = MouseButtons.Left Then
            ActivateFormByHandle(CType(item.Tag, Form))
        End If
    End Sub
    Private Function CountText(count As Integer) As String
        Dim MainText As String = "Select Window"
        If count > 0 Then
            Return MainText & " (" & count & ")"
        Else
            Return MainText
        End If
    End Function
End Class

File: AssetManager/UserInterface/CustomControls/WindowList/WindowList.cs
using AssetManager.UserInterface.Forms.Sibi;
using System;
using System.Windows.Forms;

namespace AssetManager.UserInterface.CustomControls
{
    public class WindowList : IDisposable
    {
        #region "Fields"

        private ToolStripDropDownButton dropDownControl = new ToolStripDropDownButton();
        private bool dropDownOpen = false;
        private ExtendedForm parentForm;
        private IWindowList windowListForm;
        private bool hasParentMenu = false;

        #endregion "Fields"

        #region "Constructors"

        public WindowList(ExtendedForm parentForm)
        {
            windowListForm = parentForm;
            windowListForm.ChildAdded += WindowListForm_ChildAdded;
            windowListForm.ChildRemoved += WindowListForm_ChildRemoved;
            this.parentForm = parentForm;
        }

        #endregion "Constructors"

        #region "Methods"

        public void InsertWindowList(OneClickToolStrip targetToolStrip)
        {
            InitializeDropDownButton(targetToolStrip);
            SetVisibility();
        }

        private void AddParentMenu()
        {
            if (parentForm.ParentForm != null)
            {
                var parentDropDown = NewMenuItem(parentForm.ParentForm);
                parentDropDown.Text = "[Parent] " + parentDropDown.Text;
                parentDropDown.ToolTipText = "Parent Form";
                dropDownControl.DropDownItems.Insert(0, parentDropDown);
                dropDownControl.DropDownItems.Add(new ToolStripSeparator());
                hasParentMenu = true;
            }
        }

        /// <summary>
        /// Adds a menu item for the specified child to the specified parent menu item.
        /// </summary>
        /// <param name="parent"></param>
        /// <param name="child"></param>
        /// <param name="targetMenu"></param>
        private void AddChildMenu(ExtendedForm parent, ExtendedForm child, ToolStripItemCollection targetMenu)
        {
            // If this window list parent is the direct parent of the child,
            // add the child menu item to the top/root menu.
            if (parent == this.parentForm)
            {
                // Always add the Sibi form to the top of the menu,
                // all others can go below.
                if (child is SibiMainForm)
                {
                    targetMenu.Insert(0, NewMenuItem(child));
                }
                else
                {
                    targetMenu.Add(NewMenuItem(child));
                }
            }
            else
            {
                // If this window list parent is not the direct parent of the child,
                // iterate and recurse through all the menu items until we find the
                // item which targets the parent, and add to or start a new sub menu.
                foreach (var item in targetMenu)
                {
                    // Ignore seperators.
                    if (item is OnlineStatusMenuItem)
                    {
                        var statusItem = (OnlineStatusMenuItem)item;

                        // Add the child menu item to the sub menu of the matching parent item.
                        if (statusItem.TargetForm == parent)
                        {
                            statusItem.DropDownItems.Add(NewMenuItem(child));
                        }

                        // Recurse with sub menu items.
                        if (statusItem.HasDropDownItems)
                            AddChildMenu(parent, child, statusItem.DropDownItems);
                    }
                }
            }
        }

        /// <summary>
        /// Removes the menu item for the specified child form.
        /// </summary>
        /// <param name="child"></param>
        /// <param name="targetMenu"></param>
        private void RemoveChildMenu(ExtendedForm child, ToolStripItemCollection targetMenu)
        {
            // Must use a regular 'for' block because we are going to modify the collection.
            // Iterate and recurse all menu items and remove any items that match the specified child form.
            for (int i = 0; i < targetMenu.Count; i++)
            {
                // Ignore seperators.
                if (targetMenu[i] is OnlineStatusMenuItem)
                {
                    OnlineStatusMenuItem item = (OnlineStatusMenuItem)targetMenu[i];
                    if (item.TargetForm == child)
                    {
                        DisposeDropDownItem(item);
                    }

                    if (item.HasDropDownItems)
                        RemoveChildMenu(child, item.DropDownItems);
                }
            }
        }

        private string CountText(int count)
        {
            string MainText = "Select Window";
            if (count > 0)
            {
                return MainText + " (" + count + ")";
            }
            else
            {
                return MainText;
            }
        }

        private void DisposeDropDownItem(ToolStripItem item)
        {
            // PROBLEM:
            // Because we are using the MouseUp event from the menu items,
            // and also removing/disposing those items as a result of that event,
            // the top level drop down button UI tends to hang and cause strange 
            // behaviour. The problem is particularly apparent after the last 
            // menu item has been removed. Small rectangles left in the cornor 
            // of the screen, and forms spontaneously popping infront of the 
            // active form are two of the most annoying problems.
            // 
            // REASON (maybe):
            // After much digging and testing, I've managed to narrow it down
            // to stuck/corrupt mouse event messages. These screwed up messages
            // tend to fire on other controls and cause unintended UI changes.
            // My best guess is that the item is disposed before the WM_LBUTTONUP
            // is actually received and processed by the menu control.
            //
            // I tried to find a decent logical way around it, found the following
            // to work the best.
            //
            // WORKAROUND:
            // We selectively call the PerfomClick method on menu items just
            // before they are removed and disposed. This seems to clear out
            // any remaining mouse button messages and allows the items to go
            // away cleanly and quietly. The only reason we don't call it on
            // every disposal is because it causes the top level menu to close,
            // and this is not desirable if users want to manage multiple items
            // without having to reopen the menu after every click.

            // If the item is not visible, always perform a click.
            if (!item.Visible)
            {
                item.PerformClick();
            }
            else
            {
                // If the item is the last item in a menu containing a fixed parent item, perform a click.
                // (Parent item + Separator + last item = 3)
                if (dropDownControl.DropDownItems.Count == 3 && hasParentMenu)
                {
                    item.PerformClick();
                }
                // If the item is the last item, regardless of a fixed parent item, perform a click.
                else if (dropDownControl.DropDownItems.Count == 1)
                {
                    item.PerformClick();
                }
            }

            // Do the usual clean up and disposal stuff.
            if (dropDownControl.DropDownItems.Count < 1)
            {
                dropDownControl.DropDownItems.Clear();
            }
            else
            {
                item.Visible = false;
                dropDownControl.DropDownItems.Remove(item);
            }

            item.MouseUp -= ItemClicked;
            item.Dispose();
        }

        private void DisposeAllDropDownItems()
        {
            for (int i = 0; i < dropDownControl.DropDownItems.Count; i++)
            {
                dropDownControl.DropDownItems[i].MouseUp -= ItemClicked;
                dropDownControl.DropDownItems[i].Dispose();
            }
            dropDownControl.DropDownItems.Clear();
        }

        private void DropDownControl_DropDownClosed(object sender, EventArgs e)
        {
            dropDownOpen = false;
        }

        private void DropDownControl_DropDownOpened(object sender, EventArgs e)
        {
            dropDownOpen = true;
        }

        private void InitializeDropDownButton(OneClickToolStrip targetToolStrip)
        {
            dropDownControl.Visible = false;
            dropDownControl.Font = targetToolStrip.Font;
            dropDownControl.Text = "Select Window";
            dropDownControl.Name = "WindowList";
            dropDownControl.Image = Properties.Resources.CascadeIcon;
            dropDownControl.DropDownClosed += DropDownControl_DropDownClosed;
            dropDownControl.DropDownOpened += DropDownControl_DropDownOpened;
            AddParentMenu();
            targetToolStrip.Items.Insert(targetToolStrip.Items.Count, dropDownControl);
        }

        private void ItemClicked(object sender, MouseEventArgs e)
        {
            var item = (OnlineStatusMenuItem)sender;
            var frm = item.TargetForm;
            if (e.Button == MouseButtons.Right)
            {
                if ((frm != parentForm.ParentForm))
                {
                    frm.Close();
                }
            }
            else if (e.Button == MouseButtons.Left)
            {
                if (!frm.IsDisposed)
                {
                    item.PerformClick();
                    frm.RestoreWindow();
                }
                else
                {
                    DisposeDropDownItem(item);
                }
            }
            else if (e.Button == MouseButtons.Middle)
            {
                if (!frm.IsDisposed)
                {
                    frm.WindowState = FormWindowState.Minimized;
                    parentForm.Focus();
                }
                else
                {
                    DisposeDropDownItem(item);
                }
            }
            SetVisibility();
        }

        private OnlineStatusMenuItem NewMenuItem(ExtendedForm frm)
        {
            var newitem = new OnlineStatusMenuItem();
            newitem.Name = frm.Name;
            newitem.Font = dropDownControl.Font;
            newitem.Text = frm.Text;
            newitem.Image = frm.Icon.ToBitmap();
            newitem.TargetForm = frm;

            if (frm is IOnlineStatus)
            {
                newitem.SetOnlineStatusInterface((IOnlineStatus)frm);
            }
            newitem.ToolTipText = "Right-Click to close. Middle-Click to minimize.";
            newitem.MouseUp += ItemClicked;
            return newitem;
        }

        private void SetVisibility()
        {
            if (dropDownControl.DropDownItems.Count > 0)
            {
                dropDownControl.Visible = true;
            }
            else
            {
                dropDownControl.Visible = false;
            }
        }

        private void WindowListForm_ChildAdded(object sender, ExtendedForm e)
        {
            AddChildMenu((ExtendedForm)sender, e, dropDownControl.DropDownItems);
            dropDownControl.Text = CountText(windowListForm.ChildFormCount());
            SetVisibility();
        }

        private void WindowListForm_ChildRemoved(object sender, ExtendedForm e)
        {
            RemoveChildMenu(e, dropDownControl.DropDownItems);
            dropDownControl.Text = CountText(windowListForm.ChildFormCount());
            SetVisibility();
        }

        #endregion "Methods"

        #region "IDisposable Support"

        private bool disposedValue;

        public void Dispose()
        {
            Dispose(true);
        }

        protected virtual void Dispose(bool disposing)
        {
            if (!disposedValue)
            {
                if (disposing)
                {
                    DisposeAllDropDownItems();
                    dropDownControl.Dispose();
                    windowListForm.ChildAdded -= WindowListForm_ChildAdded;
                    windowListForm.ChildRemoved -= WindowListForm_ChildRemoved;
                    dropDownControl.DropDownClosed -= DropDownControl_DropDownClosed;
                    dropDownControl.DropDownOpened -= DropDownControl_DropDownOpened;
                }
            }
            disposedValue = true;
        }

        #endregion "IDisposable Support"
    }
}